coveo-functools/coveo_functools/flex.py

```diff
@@ -0,0 +1,300 @@
+
+_ = unflex, flexcase  # mark them as used (forward compatibility vs docs)
+
+
+JSON_TYPES = (
+    str,
+    bool,
+    int,
+    float,
+    type(None),
+    dict,
+)  # list omitted to support list of custom types
+PASSTHROUGH_TYPES = {None, Any, *JSON_TYPES}
+
+T = TypeVar("T")
+
+TypeHint = Any  # :shrug:
+
+RealClass = Type[T]
+RealFunction = Callable[..., T]
+RealObject = Union[RealClass, RealFunction]
+WrappedClass = Type[T]
+WrappedFunction = Callable[..., T]
+WrappedObject = Union[RealClass, RealFunction]
+
+RAW_KEY: Final[str] = "_flexed_from_"
+
+
+@overload
+def flex() -> Callable[[RealObject], WrappedObject]:
+    ...
+
+
+@overload
+def flex(obj: None) -> Callable[[RealObject], WrappedObject]:
+    ...
+
+
+@overload
+def flex(obj: RealClass) -> WrappedClass:
+    ...
+
+
+@overload
+def flex(obj: RealFunction) -> WrappedFunction:
+    ...
+
+
+def flex(
+    obj: Optional[RealObject] = None,
+) -> Union[WrappedObject, Callable[[RealObject], WrappedObject]]:
+    """Wraps `obj` into recursive flexcase magic."""
+    if obj is not None:
+
+        """
+        Covers decorator usages without parenthesis:
+
+            @flex
+            def obj(...)
+
+            @flex
+            class C:
+                @flex
+                def obj(self, ...)
+                    ...
+
+        Also covers the complete inline usage:
+
+            f = flex(obj, ...)(**dirty_kwargs)
+
+        """
+
+        return _generate_wrapper(obj)
+
+    else:
+
+        """
+        Covers decorator usages with parenthesis:
+
+            @flex()
+            def obj(...)
+
+            @flex()
+            class C:
+
+                @flex()
+                def obj(self, ...)
+                    ...
+        """
+
+        # python's mechanic is going to call us again with the obj as the first (and only) argument to get a wrapper.
+        return flex
+
+
+def deserialize(value: Any, *, hint: TypeHint) -> Any:
+    """
+    Deserializes a value based on the provided type hint:
+
+    - If value is None, we return None. We don't validate the hint.
+    - If hint is a builtin type, we blindly return the value we were given. We don't validate the value.
+    - If hint is a custom class, the value must be a dictionary to "flex" into the class.
+    - Hint may be a `Union[T, List[T]]`, in which case the value must be a dict or a list of them.
+    """
+    if value is None:
+        return None  # nope!
+
+    # origin: like `list` for `List` or `Union` for `Optional`
+    # args: like (str, int) for Optional[str, int]
+    origin, args = _resolve_hint(hint)
+
+    if origin is Union:
+        if not {*args}.difference(PASSTHROUGH_TYPES):
+            # Unions of PASSTHROUGH_TYPES are allowed and assumed to be in the proper type already
+            return value
+
+        # implementation detail: the `_resolve_hint` function always puts the real type first.
+        target_type: TypeHint = args[0]

```

Comment:
So if you modify `_resolve_hint`, you may break this line. I understand it might be difficult to do otherwise. Still, if you are able to move these lines of code next to the `_resolve_hint` call at line `133`, it might mitigate the risk (you'd need to use `target_type = args[0] if args else Any`)

Reply:
i'll see if I can improve that little part 🤔

Reply:
Let me know if this is better now: https://github.com/coveooss/coveo-python-oss/pull/78/commits/30277047600919a386a1ef0d4a91e0385d385793